Better SEO, OpenGraph, schema.org and clean generated code

jmrplens opened the pull request:
I have made an html that generates all the information for [OpenGraph](https://ogp.me/) and also the [Schema.org](https://schema.org/) in head.html.

Right now, schema it is intended for personal web use, it would be necessary to add an option if the web is from a laboratory or group to get correct Schema.org.

[Schema test (my web).](https://validator.schema.org/#url=https%3A%2F%2Fjmrplens.github.io%2F)

[OpenGraph test (my web).](https://www.opengraph.xyz/url/https%3A%2F%2Fjmrplens.github.io/)

**UPDATE 1:**

Added HTML compression to remove blank lines generated by liquid code. Thanks to [jekyll-compress-html](https://github.com/penibelst/jekyll-compress-html).

**UPDATE 2:**

- Added: google and bing site verification
- Minor corrections

### [CHANGELOG](https://github.com/alshedivat/al-folio/pull/481#issuecomment-991958126)

rohandebsarkar commented:
**SUGGESTION:**

Adding:
```
sass:
  style: compressed
```
to `_config.yml` will also make the compiled CSS compressed.

In my testing, the `assets/css/main.css` file was reduced by about 2 KB.

jmrplens commented:
@rohandebsarkar I'll try it tomorrow and add it.

Thanks!

jmrplens commented:
Added the CSS compression suggested by @rohandebsarkar. Approximately 15% less weight is obtained.

**Other suggestion:**

I use [tinypng](https://tinypng.com/) to compress the images before uploading to the repository, I am sure that there must be some way to implement the API and that the process is carried out in the build.
[Here is some information](https://spock.rocks/tech/2016/04/02/compressing-images-with-tinypng.html), we can see it for another pull request or try to implement it in this one.

rohandebsarkar commented:
> Added the CSS compression suggested by @rohandebsarkar. Approximately 15% less weight is obtained.
> 
> **Other suggestion:**
> 
> I use [tinypng](https://tinypng.com/) to compress the images before uploading to the repository, I am sure that there must be some way to implement the API and that the process is carried out in the build. [Here is some information](https://spock.rocks/tech/2016/04/02/compressing-images-with-tinypng.html), we can see it for another pull request or try to implement it in this one.

We can rather use the [`jekyll-webp`](https://github.com/sverrirs/jekyll-webp) plugin to compress and convert the images to WebP format.

WebP images are much smaller compared to PNG and JPEG, faster to load and most of the mordern browsers recommend it.

More information about WebP: https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#webp_image

jmrplens commented:
> > Added the CSS compression suggested by @rohandebsarkar. Approximately 15% less weight is obtained.
> > **Other suggestion:**
> > I use [tinypng](https://tinypng.com/) to compress the images before uploading to the repository, I am sure that there must be some way to implement the API and that the process is carried out in the build. [Here is some information](https://spock.rocks/tech/2016/04/02/compressing-images-with-tinypng.html), we can see it for another pull request or try to implement it in this one.
> 
> We can rather use the [`jekyll-webp`](https://github.com/sverrirs/jekyll-webp) plugin to compress and convert the images to WebP format.
> 
> WebP images are much smaller compared to PNG and JPEG, faster to load and most of the mordern browsers recommend it.
> 
> More information about WebP: https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#webp_image

I just tried it but I have not been able to convert the images of the blog (my main image already had it on webp).

Have you been able to prove it?

I have also seen a problem with the compression of HTML and the code blocks, I have done several tests and with a code block like the following, it does not display well:

```liquid
{% highlight c++ linenos %}

int main(int argc, char const \*argv[])
{
    string myString;

    cout << "input a string: ";
    getline(cin, myString);
    int length = myString.length();

    char charArray = new char * [length];

    charArray = myString;
    for(int i = 0; i < length; ++i){
        cout << charArray[i] << " ";
    }

    return 0;
}

{% endhighlight %}
```

Later I will try this plugin that may work better: https://github.com/digitalsparky/jekyll-minifier

rohandebsarkar commented:
@jmrplens the `jekyll-webp` plugin is deprecated, we can rather use the [`jekyll-imagemagick`](https://gitlab.com/emmmile/jekyll-imagemagick) plugin.
I opened a PR #498 to convert all images to WebP format.

jmrplens commented:
## Final commit.

### Summary:

- Added OpenGraph and Twitter Card metadata. (file: `_includes/metadata.html`)
- Added Schema.org metadata. (file: `_includes/metadata.html`)
- Added site verification (Google and Bing). (file: `_includes/metadata.html`)
- Added new layout to parse HTML code. Loaded from default layout. (file: `_layouts/parse.html`)
- Added HTML beautifier, has conflict with code blocks. Set in `_config.yml`. (Issue: https://github.com/threedaymonk/htmlbeautifier/issues/62). (file: `_plugins/beautify.rb`).
- Added HTML minify. Set in `_config.yml`. (file: `_plugins/minify.rb`)
- Added SASS compresion in `_config.yml`.
- Small revision to update part of the code and not generate extra white spaces. It may need more changes. [Whitespace Control](https://shopify.github.io/liquid/basics/whitespace/).

alshedivat commented on _config.yml:
```diff
@@ -13,6 +13,7 @@ footer_text: >
   Powered by <a href="http://jekyllrb.com/" target="_blank">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio">al-folio</a> theme.
   Hosted by <a href="https://pages.github.com/" target="_blank">GitHub Pages</a>.
   Photos from <a href="https://unsplash.com" target="_blank">Unsplash</a>.
+keywords: al-folio, github, etc
```
```suggestion
keywords: jekyll, jekyll-theme, academic-website, portfolio-website  # add your own keywords or leave empty
```

alshedivat commented on _config.yml:
```diff
@@ -73,6 +75,8 @@ contact_note: >
   You can even add a little note about which of these is the best way to reach you.
 
 google_analytics: UA-XXXXXXXXX # out your google-analytics code
+google_site_verification: # out your google-site-verification code (Google Search Console)
+bing_site_verification: # out your bing-site-verification code (Bing Webmaster)
 panelbear_analytics: XXXXXXXXX # panelbear analytics site ID
```
let's improve defaults of this entire block:
```suggestion
google_analytics:  # your google-analytics ID (format: UA-XXXXXXXXX)
google_site_verification:  # your google-site-verification ID (Google Search Console)
bing_site_verification:  # out your bing-site-verification ID (Bing Webmaster)
panelbear_analytics:  # panelbear analytics site ID (format: XXXXXXXXX)
```

alshedivat commented on _config.yml:
```diff
@@ -158,10 +162,27 @@ defaults:
     values:
       sitemap:         false
-
+    
```
unnecessary extra spaces
```suggestion

```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<!-- Standard metadata -->
+{%- if site.enable_google_verification -%}
+<meta name="google-site-verification" content="{{ site.google_site_verification }}" />
+{%- endif -%}
+{%- if site.enable_bing_verification -%}
+<meta name="msvalidate.01" content="{{ site.bing_site_verification }}" />
+{%- endif -%}
```
question: wouldn't adding `-` to both sides of `{%-` and `-%}` result in having both `<meta />` tags on the same line? (this is a cool pattern btw, didn't know about it)

jmrplens replied:
I have had to nest conditionals to keep the two labels on the same line, also when only one of them is activated.

```liquid
{%- if site.enable_google_verification or site.enable_bing_verification %}
    <!-- Website verification -->
    {% if site.enable_google_verification -%}
    <meta name="google-site-verification" content="{{ site.google_site_verification }}" />
    {%- endif -%}
    {% if site.enable_bing_verification -%}
    <meta name="msvalidate.01" content="{{ site.bing_site_verification }}" />
    {%- endif -%}
{%- endif %}
```

I have also left the HTML comments inside the conditionals, so that they are only printed if the site verification, opengraph, etc. is generated.

PR: https://github.com/alshedivat/al-folio/pull/481/commits/5dfa4447ef746e3dd0418e989c32f52bb20813bb

The final result with data is:
<img width="1475" alt="Screenshot of Google Chrome (22-12-21, 10-47-55)" src="https://user-images.githubusercontent.com/28966312/147072905-be3794e4-9f63-44b9-b97b-d7c74da4cb4e.png">

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta name="author" content="{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}" />
+<meta name="description" content="{%- if page.description -%}{{ page.description }}{%- else -%}{{ site.description }}{%- endif -%}" />
+{% if page.keywords or site.keywords %}<meta name="keywords" content="{%- if page.keywords -%}{{ page.keywords }}{%- else -%}{{ site.keywords }}{%- endif -%}" />{% endif %}
+<!-- OpenGraph -->
```
```suggestion

<!-- OpenGraph -->
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta property="og:description" content="{%- if page.description -%}{{ page.description }}{%- else -%}{{ site.description }}{%- endif -%}" />
+{% if page.og_image or site.og_image %}<meta property="og:image" content="{%- if page.og_image -%}{{ page.og_image }}{%- else -%}{{ site.og_image }}{%- endif -%}" />{% endif %}
+<meta property="og:locale" content="{{ site.lang }}" />
+<!-- Twitter card -->
```
```suggestion

<!-- Twitter card -->
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta name="twitter:card" content="summary" />
+<meta name="twitter:title" content="{%- if page.title -%}{{ page.title }}{%- else -%}{{ site.title }}{%- endif -%}" />
+<meta name="twitter:description" content="{%- if page.description -%}{{ page.description }}{%- else -%}{{ site.description }}{%- endif -%}" />
+{% if page.og_image or site.og_image %}<meta name="twitter:image" content="{%- if page.og_image -%}{{ page.og_image }}{%- else -%}{{ site.og_image }}{%- endif -%}" />{% endif %}
```
```suggestion
{% if page.og_image or site.og_image -%}
<meta name="twitter:image" content="{%- if page.og_image -%}{{ page.og_image }}{%- else -%}{{ site.og_image }}{%- endif -%}" />
{%- endif %}
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta name="twitter:site" content="@{{ site.twitter_username }}" />
+<meta name="twitter:creator" content="@{{ site.twitter_username }}" />
+{%- endif -%}
+{%- endif -%}
```
what does this `endif` correspond to?

jmrplens replied:
To the OpenGraph condition. More clear in PR https://github.com/alshedivat/al-folio/pull/481/commits/49924aea02900c6405001a875b9df098d3ba9a5d

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+<meta http-equiv="X-UA-Compatible" content="IE=edge">
+<title>
+{%- if site.title == "blank" -%}
+{%- capture title -%}{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}{%- endcapture -%}
+{%- else -%}
+{%- capture title -%}{{ site.title }}{%- endcapture -%}
+{%- endif -%}
+{%- if page.title != "blank" and page.url != "/" -%}
+{{ title }} | {{ page.title }}
+{%- else -%}
+{{ title }}
+{%- endif -%}
```
do you mind correctly indenting this block of code with 4 spaces (no tabs) for better readability? (same request for the code below)

jmrplens replied:
Solved in https://github.com/alshedivat/al-folio/pull/481/commits/9648f1bcfff8c0f698118436f1a4c8f5e8f8d518

Source code example with full data:
<img width="1352" alt="Screenshot of Google Chrome (22-12-21, 09-46-06)" src="https://user-images.githubusercontent.com/28966312/147064567-3a1f5cbe-e0ef-4b46-bfd3-0ce02fc73614.png">

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+</title>
+<meta name="author" content="{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}" />
+<meta name="description" content="{%- if page.description -%}{{ page.description }}{%- else -%}{{ site.description }}{%- endif -%}" />
+{% if page.keywords or site.keywords %}<meta name="keywords" content="{%- if page.keywords -%}{{ page.keywords }}{%- else -%}{{ site.keywords }}{%- endif -%}" />{% endif %}
```
```suggestion
{% if page.keywords or site.keywords -%}
    <meta name="keywords" content="{%- if page.keywords -%}{{ page.keywords }}{%- else -%}{{ site.keywords }}{%- endif -%}" />
{%- endif %}
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta property="og:title" content="{%- if page.title -%}{{ site.title }} | {{ page.title }}{%- else -%}{{ site.title }}{%- endif -%}" />
+<meta property="og:url" content="{{ page.url | prepend: site.baseurl | prepend: site.url | remove_first: 'index.html' }}" />
+<meta property="og:description" content="{%- if page.description -%}{{ page.description }}{%- else -%}{{ site.description }}{%- endif -%}" />
+{% if page.og_image or site.og_image %}<meta property="og:image" content="{%- if page.og_image -%}{{ page.og_image }}{%- else -%}{{ site.og_image }}{%- endif -%}" />{% endif %}
```
```suggestion
{% if page.og_image or site.og_image -%}
    <meta property="og:image" content="{%- if page.og_image -%}{{ page.og_image }}{%- else -%}{{ site.og_image }}{%- endif -%}" />
{%- endif %}
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<meta name="twitter:creator" content="@{{ site.twitter_username }}" />
+{%- endif -%}
+{%- endif -%}
+<!-- Schema.org -->
```
```suggestion

<!-- Schema.org -->
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+{%- endif -%}
+<!-- Schema.org -->
+{%- if site.serve_schema_org -%}
+	<!-- Social links generator for "sameAs schema" -->
```
do you mind converting indentation tabs to 4 spaces each?

jmrplens replied:
Solved in https://github.com/alshedivat/al-folio/pull/481/commits/9648f1bcfff8c0f698118436f1a4c8f5e8f8d518

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,142 @@
+<!-- Schema.org -->
+{%- if site.serve_schema_org -%}
+	<!-- Social links generator for "sameAs schema" -->
+	{%- assign sameaslinks = "" | split: "," -%}
```
the code below looks pretty verbose... we can definitely leave this to a later PR, but this would probably work better if we nest all these profile ids in `_config.yml` under `profiles:` and then use a for loop here iterate over all of them. what do you think?

jmrplens replied:
Good idea. But in _config.yml, to keep it simple, just enter the username or id, a script that builds the urls would be necessary.
Have you thought of something specific?

It only occurs to me, a script would be made to generate all the social urls, and use that script for both the schema and the social icon bar. See if someone else comes up with another idea.

alshedivat replied:
yes, you are totally right that it would still require case-by-case URL generation... never mind  then, let's keep it as is.

alshedivat commented on _includes/news.html:
```diff
@@ -1,24 +1,24 @@
 <div class="news">
   <h2>news</h2>
-  {% if site.news  %}
+  {%- if site.news  -%} 
```
wouldn't adding `-` on both sides place `<h2>news</h2>` and `<div class="table-responsive">` on the same line? (same question about other places where `{%-` and `-%}` where added)

jmrplens replied:
Yes, just add the one on the right, not both.

Result:
```html
          <!-- News -->          
          <div class="news">
            <h2>news</h2>
            <div class="table-responsive">
              <table class="table table-sm table-borderless"> 
                <tr>
                  <th scope="row">Jan 15, 2016</th>
                  <td>
                    A simple inline announcement with Markdown emoji! <img class="emoji" title=":sparkles:" alt=":sparkles:" src="https://github.githubassets.com/images/icons/emoji/unicode/2728.png" height="20" width="20"> <img class="emoji" title=":smile:" alt=":smile:" src="https://github.githubassets.com/images/icons/emoji/unicode/1f604.png" height="20" width="20">
 
                  </td>
                </tr> 
                <tr>
                  <th scope="row">Nov 7, 2015</th>
                  <td>
                    <a class="news-title" href="/news/announcement_2/">A long announcement with details</a> 
                  </td>
                </tr> 
                <tr>
                  <th scope="row">Oct 22, 2015</th>
                  <td>
                    A simple inline announcement.
 
                  </td>
                </tr> 
              </table>
            </div> 
          </div>
```

I have made more changes so that the HTML output is correct, with 2 spaces (I leave information below). PR: https://github.com/alshedivat/al-folio/pull/481/commits/3808d418d5890ebfc3050a71754d69e04429c103
In Liquid when including html or loading content does not respect indents. Each file must be indented separately.

[Google HTML Formatting](https://developers.google.com/style/html-formatting)
[W3 HTML Style](https://www.w3schools.com/html/html5_syntax.asp)

alshedivat commented on _includes/social.html:
```diff
@@ -1,20 +1,19 @@
+{%- if site.gitlab_username -%}<a href="https://gitlab.com/{{ site.gitlab_username }}" title="GitLab"><i class="fab fa-gitlab"></i></a>{%- endif -%}
+{%- if site.dblp_url -%}<a href="{{ site.dblp_url }}" title="DBLP"><i class="ai ai-dblp"></i></a>{%- endif -%}
+{%- if site.stackoverflow_id -%}<a href="https://stackoverflow.com/users/{{ site.stackoverflow_id }}" title="Stackoverflow"><i class="fab fa-stack-overflow"></i></a>{%- endif -%}
+{%- if site.rss_icon -%}<a href="{{ site.baseurl | prepend: site.url }}/feed.xml" title="RSS Feed"><i class="fas fa-rss-square"></i></a>{%- endif -%}
```
same question as above: wouldn't adding `-` to all `{%` and `%}` place all these anchors on the same line in the generated code? I'd prefer placing each on a separate line, if possible.

jmrplens replied:
PR: https://github.com/alshedivat/al-folio/pull/481/commits/bd1d4f078c3ef78fb007538a35d8a81cb2771b4f

### Social in NavBar
<img width="1439" alt="Screenshot of Google Chrome (22-12-21, 14-58-45)" src="https://user-images.githubusercontent.com/28966312/147104265-df8d299f-b24e-4efc-be71-31394aab5010.png">

### Social in About (bottom)
<img width="1544" alt="Screenshot of Google Chrome (22-12-21, 15-03-39)" src="https://user-images.githubusercontent.com/28966312/147104821-bdaccbf8-bffe-4087-8e8a-02f5b33ba67b.png">

alshedivat commented on _layouts/parse.html:
```diff
@@ -0,0 +1,9 @@
+		{%- if forloop.first -%}{%- continue -%}{%- endif -%}{%- capture _comment_outside -%}{%- if _carry -%}{{ _comments.first }}{%- endif -%}{{ _comment_before }}{%- endcapture -%}{%- capture _comment -%}{%- unless _carry -%}{{ _comments.first }}{%- endunless -%}{{ _comment_outside | split: _comments.last | first }}{%- if _comment_outside contains _comments.last -%}{{ _comments.last }}{%- assign _ca ...
+	{%- endfor -%}
+{%- endif -%}
+{{ content | beautify | minify }}
```
```suggestion
{{ content | beautify | minify }}

```

alshedivat commented on _layouts/parse.html:
```diff
@@ -0,0 +1,9 @@
+{%- if site.remove_HTML_comments -%}
+	{%- assign _comments = "<!-- -->" | split: " " -%}
+	{%- capture _comment_befores -%}.{{ content }}{%- endcapture -%}
```
is the `.` before `{{ content }}` intentional?

jmrplens replied:
It's a mistake. Solved in PR: https://github.com/alshedivat/al-folio/pull/481/commits/3cb3ebe13f6e5d89f92b21ab7d7b99a21c89e635

alshedivat commented on _layouts/parse.html:
```diff
@@ -0,0 +1,9 @@
+	{%- capture _comment_befores -%}.{{ content }}{%- endcapture -%}
+	{%- assign _comment_befores = _comment_befores | split: _comments.first -%}
+	{%- for _comment_before in _comment_befores -%}
+		{%- if forloop.first -%}{%- continue -%}{%- endif -%}{%- capture _comment_outside -%}{%- if _carry -%}{{ _comments.first }}{%- endif -%}{{ _comment_before }}{%- endcapture -%}{%- capture _comment -%}{%- unless _carry -%}{{ _comments.first }}{%- endunless -%}{{ _comment_outside | split: _comments.last | first }}{%- if _comment_outside contains _comments.last -%}{{ _comments.last }}{%- assign _ca ...
```
I would split this logic into multiple lines for readability, if possible. it shouldn't introduce new whitespaces b/c of `-` pattern used, right?

alshedivat commented on _plugins/beautify.rb:
```diff
@@ -0,0 +1,26 @@
+require 'htmlbeautifier'
```
```suggestion
require 'htmlbeautifier'

```

alshedivat commented on _plugins/beautify.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+end
+
+Liquid::Template.register_filter(Jekyll::StripHTML)
```
```suggestion
Liquid::Template.register_filter(Jekyll::StripHTML)

```

alshedivat commented on _plugins/minify.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+end
+
+Liquid::Template.register_filter(Jekyll::MinifyFilter)
```
```suggestion
Liquid::Template.register_filter(Jekyll::MinifyFilter)

```

alshedivat reviewed (commented):
this is a tremendous amount of work, thank you for sending this PR!! 🙏

LGTM overall. I left a bunch of inline comments below, most of which are very minor stylistic suggestions. it would be great if you could take a look and address some of them.

jmrplens commented:
> this is a tremendous amount of work, thank you for sending this PR!! 🙏
> 
> LGTM overall. I left a bunch of inline comments below, most of which are very minor stylistic suggestions. it would be great if you could take a look and address some of them.

Thanks!
I will try to solve all the ideas and doubts soon. I'm doing it bit by bit :)

jmrplens commented:
I have been editing while reviewing the generated HTML code (PR:https://github.com/alshedivat/al-folio/pull/481/commits/3808d418d5890ebfc3050a71754d69e04429c103), I have been able to fix everything except the content block (different for each page), it needs more time, but I think it is quite good.

[Google HTML Formatting](https://developers.google.com/style/html-formatting)
[W3 HTML Style](https://www.w3schools.com/html/html5_syntax.asp)

Some examples with personal data:

### HEAD
<img width="1269" alt="Screenshot of Google Chrome (22-12-21, 14-22-37)" src="https://user-images.githubusercontent.com/28966312/147099295-177f12bb-e002-4d8b-b78d-0e8182d6da5a.png">

### BODY/HEADER
<img width="1117" alt="Screenshot of Google Chrome (22-12-21, 14-23-31)" src="https://user-images.githubusercontent.com/28966312/147099434-57ede299-ed20-4dd5-911d-50a828820262.png">

### BODY/FOOTER & SCRIPTS
<img width="1051" alt="Screenshot of Google Chrome (22-12-21, 14-24-11)" src="https://user-images.githubusercontent.com/28966312/147099525-1fdf3322-1cb4-430a-8e6d-0cfd45a06111.png">

jmrplens commented:
In PR: https://github.com/alshedivat/al-folio/pull/481/commits/5438a9a75fcf8c0dae9e5b1c0e6351c8bb470462 

From this:
<img width="667" alt="Screenshot of Sublime Text (23-12-21, 23-25-14)" src="https://user-images.githubusercontent.com/28966312/147297611-d86229f1-8b22-458a-82ea-21d4e9a6887f.png">

To:
<img width="684" alt="Screenshot of Sublime Text (23-12-21, 23-27-33)" src="https://user-images.githubusercontent.com/28966312/147297623-c3ae3e0f-af57-4a25-9b0a-4e39113d7d33.png">

jmrplens commented:
Beautify bib generator. PR: https://github.com/alshedivat/al-folio/pull/481/commits/3fe1f3dd6fb9ff57fcec7230b3f180c796fc2d96

From this:
<img width="1079" alt="Screenshot of Sublime Text (24-12-21, 03-12-02)" src="https://user-images.githubusercontent.com/28966312/147307295-bfc0cc3b-a59e-40e7-8fcf-ff1d43b4dc39.png">

To:
<img width="797" alt="Screenshot of Sublime Text (24-12-21, 03-07-00)" src="https://user-images.githubusercontent.com/28966312/147307107-e5c77b2a-0d66-453a-85cf-01cadf89d86b.png">

jmrplens commented:
I have made a new change. On the 'about' page, a subtitle is shown that may have HTML code (links, table, etc), this type of subtitle may be erroneously displayed in search engines. It is better to separate and have a subtitle variable and a description variable for the meta tag.
In the event that the subtitle does not exist, the description will be used.

PR's: https://github.com/alshedivat/al-folio/pull/481/commits/ea2a3963f25e096814abfde1718e7f8f5ea38d3f https://github.com/alshedivat/al-folio/pull/481/commits/f23793b366dd2874aea61b254ec792437ab6957c https://github.com/alshedivat/al-folio/pull/481/commits/964b0eb741054614d5f9cd3cef4fea3532bb44a3

alshedivat commented on _includes/projects.html:
```diff
@@ -1,34 +1,36 @@
-          <div class="github-icon">
-            <div class="icon" data-toggle="tooltip" title="Code Repository">
-              <a href="{{ project.github }}"><i class="fab fa-github gh-icon"></i></a>
+
```
sorry, what happened with indentation in this file? seems like a bunch of extra white spaces got added a the begging of each line. let's fix this before merge.

alshedivat commented on _layouts/about.html:
```diff
@@ -2,52 +2,57 @@
-    {% endif %}
-  </article>
+<!-- about.html -->
+      <div class="post">
```
same comment about the extra spaces in this file...

alshedivat commented on _layouts/bib.html:
```diff
@@ -1,151 +1,158 @@
 ---
 ---
+<!-- _layouts/bib.html -->
```
same comment about the extra spaces in this file...

alshedivat commented on _layouts/page.html:
```diff
@@ -1,15 +1,16 @@
 layout: default
 ---
-<div class="post">
+<!-- page.html -->
```
same comment about the extra spaces in this file...

alshedivat commented on _pages/about.md:
```diff
@@ -2,7 +2,9 @@
 title: about
 permalink: /
-description: <a href="#">Affiliations</a>. Address. Contacts. Moto. Etc.
+description: This is a description that is used in meta tags and will show up in search engine results.
+
```
```suggestion
```
let's remove description from the about page and to make it intentionally fall back to `site.description`. you already have the fall back implemented in `_includes/metadata.html`: 
https://github.com/alshedivat/al-folio/blob/964b0eb741054614d5f9cd3cef4fea3532bb44a3/_includes/metadata.html#L28

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,158 @@
+        },
+        "url":"{{ page.url | prepend: site.baseurl | prepend: site.url }}",
+        "@type":"WebSite",
+        {% if page.description -%}
+        "description":"{{ page.description }}",
+        {% else if site.description -%}
+        "description":"{{ site.description }}",
+        {%- endif -%}
```
minor: let's inline this code block:
```suggestion
        "description": "{%- if page.description -%}{{ page.description }}{%- else if site.description -%}{{ site.description }}{%- endif -%}",
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,158 @@
+          "@type": "Person",
+          "name": "{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}"
+        },
+        "url":"{{ page.url | prepend: site.baseurl | prepend: site.url }}",
+        "@type":"WebSite",
```
```suggestion
        "url": "{{ page.url | prepend: site.baseurl | prepend: site.url }}",
        "@type": "WebSite",
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,158 @@
+        {% else if site.description -%}
+        "description":"{{ site.description }}",
+        {%- endif -%}
+        "headline":"{%- if page.title -%}{{ page.title }}{%- else -%}{{ site.title }}{%- endif -%}",
+        {% if sameaslinks != blank -%}
+        "sameAs":{{ sameaslinks }}, 
+        {%- endif %}
+        "name":"{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}",
+        "@context":"https://schema.org"
```
```suggestion
        "headline": "{%- if page.title -%}{{ page.title }}{%- else -%}{{ site.title }}{%- endif -%}",
        {% if sameaslinks != blank -%}
        "sameAs": {{ sameaslinks }}, 
        {%- endif %}
        "name": "{{ site.first_name }} {{ site.middle_name }} {{ site.last_name }}",
        "@context": "https://schema.org"
```

alshedivat commented on _includes/metadata.html:
```diff
@@ -0,0 +1,158 @@
+        "@context":"https://schema.org"
+      }
+    </script>
+{%- endif %}
```
```suggestion
{%- endif %}

```

alshedivat commented on _layouts/about.html:
```diff
@@ -2,52 +2,57 @@
+          <h1 class="post-title">
+           {% if site.title == "blank" -%}<span class="font-weight-bold">{{ site.first_name }}</span> {{ site.middle_name }} {{ site.last_name }}{%- else -%}{{ site.title }}{%- endif %}
+          </h1>
+          <p class="desc">{%- if page.subtitle -%}{{ page.subtitle }}{%- else -%}{{ page.description }}{%- endif -%}</p>
```
```suggestion
          <p class="desc">{{ page.subtitle }}</p>
```
let's remove fall back to `page.description` here. about page shouldn't have description field specified, so that the meta tag uses `site.description` for that page. you already have 
fall back to site description implemented in `_includes/metadata.html`:
https://github.com/alshedivat/al-folio/blob/964b0eb741054614d5f9cd3cef4fea3532bb44a3/_includes/metadata.html#L28

alshedivat reviewed (commented):
LGTM. thanks again for this PR!

see below just a few minor stylistic comments. happy to merge right after they are addressed.

jmrplens commented:
@alshedivat About extra spaces.

The liquid code does not maintain the indents when a file is included, the spaces of the file that includes it must be inserted in the external file so that the HTML code is indented correctly.

For example in `default.html` when include `head.html` has 4 indentation spaces:
https://github.com/alshedivat/al-folio/blob/302c8e8c74dcb5120544b0df158f7901ef862850/_layouts/default.html#L12

In `head.html` the 4 indentation spaces are kept:

https://github.com/alshedivat/al-folio/blob/302c8e8c74dcb5120544b0df158f7901ef862850/_includes/head.html#L3-L18

I have not found another way to maintain the correct indentation. The indentation must be inherited when the external file is included.

Sorry about my English, is the explanation understood?

alshedivat commented:
@jmrplens, thanks for making changes and explaining the reasoning behind indentation. I searched a bit and it looks like there is a related open issue in liquid: https://github.com/Shopify/liquid/issues/925.

I'm trying to understand this better: html beautifier is supposed to take care of white space issues and indentation, correct? are you trying to make html code look nice manually without having to use html beautifier plugin because of the issue with code blocks? or are you trying to achieve something else with these manual adjustments?

relatedly, have you seen [jekyll-tidy](https://github.com/apsislabs/jekyll-tidy)? would it work better?

jmrplens commented:
> @jmrplens, thanks for making changes and explaining the reasoning behind indentation. I searched a bit and it looks like there is a related open issue in liquid: https://github.com/Shopify/liquid/issues/925.
> 
> 
> 
> I'm trying to understand this better: html beautifier is supposed to take care of white space issues and indentation, correct? are you trying to make html code look nice manually without having to use html beautifier plugin because of the issue with code blocks? or are you trying to achieve something else with these manual adjustments?

The idea is to reduce the white spaces to reduce the weight of the files (without reaching the extreme, minify) and also with a good indentation improves the analysis carried out by search engines.
As there is the problem with the code blocks, I don't see a good idea to use the htmlbeautifier in a general way, only for those that do not show code blocks on the web.

> relatedly, have you seen [jekyll-tidy](https://github.com/apsislabs/jekyll-tidy)? would it work better?

Jekyll-Tidy use htmlbeautifier, have the same problem :(

jmrplens commented:
anyway, we can undo the extra indentation to improve readability, it's not critical.

alshedivat commented:
@jmrplens, no worries, let's keep the extra indentation that you added (I appreciate the effort!). I'll go ahead an merge.

we should look into getting html beautifier working correctly. once the issue with code blocks is fixed, we should revert all the manual indentation and start using `jekyll-tidy` instead.
